semantic_conventions/http-common.yaml

```
@@ -5,10 +5,54 @@ groups:
     prefix: http
     attributes:
       - id: request.method
-        type: string
+        type:
+          allow_custom_values: true
+          members:
+            - id: connect
+              value: "CONNECT"
+              brief: 'CONNECT method.'
+            - id: delete
+              value: "DELETE"
+              brief: 'DELETE method.'
+            - id: get
+              value: "GET"
+              brief: 'GET method.'
+            - id: head
+              value: "HEAD"
+              brief: 'HEAD method.'
+            - id: options
+              value: "OPTIONS"
+              brief: 'OPTIONS method.'
+            - id: patch
+              value: "PATCH"
+              brief: 'PATCH method.'
+            - id: post
+              value: "POST"
+              brief: 'POST method.'
+            - id: put
+              value: "PUT"
+              brief: 'PUT method.'
+            - id: trace
+              value: "TRACE"
+              brief: 'TRACE method.'
+            - id: other
+              value: "other"
+              brief: 'Any HTTP method that the instrumentation has no prior knowledge of.'
         requirement_level: required
         brief: 'HTTP request method.'
         examples: ["GET", "POST", "HEAD"]
+        note: |
+          HTTP request method SHOULD be one of the methods defined in [RFC9110](https://www.rfc-editor.org/rfc/rfc9110.html#name-methods)
+          or the PATCH method defined in [RFC5789](https://www.rfc-editor.org/rfc/rfc5789.html).
+
+          General-purpose HTTP instrumentations SHOULD allow overriding the list of known HTTP methods with OTEL_INSTRUMENTATION_HTTP_KNOWN_METHODS environment variable,
+          containing a comma-separated list of case-sensitive known HTTP methods.
+
+          If the HTTP request method is not known to instrumentation, it MUST set the `http.request.method` attribute to `other` and, if it reports spans, MUST
+          populate the exact method passed in the request line on `http.request.method_original` span attribute.
```

```suggestion
          If the HTTP request method is not known to instrumentation, it MUST set the `http.request.method` attribute to `other` and, if it reports spans, MUST
          populate the exact method passed in the request line on `http.request.method_original` span attribute.

          General-purpose HTTP instrumentations MUST provide some way for instrumentation users to add more known methos. They SHOULD implement this by allowing to overriding the list of known HTTP methods with an environment variable environment variable `OTEL_INSTRUMENTATION_HTTP_KNOWN_METHODS`,
          containing a comma-separated list of case-sensitive known HTTP methods.
```

Changes:
1. Re-order because the introduction of what a "known" method means should come before the paragraph that explains what to do with unknown methods
2. Add MUST-requirement to provide "some way" of adding more methods, make the previous SHOULD requirement the recommended way to implement this.

Point 2, that adding additional methods is a MUST, was one of my two main concerns in my previous review, the standard way is a nice-to-have relative to that.

---

NB: I did not change it in my suggestion, but it seems more practical to have OTEL_INSTRUMENTATION_HTTP_ADDITIONAL_KNOWN_METHODS (note "ADDITIONAL") than a full override that would also have to contain GET,POST, etc.

---

I really doubt that env var is the perfect solution and expect otel to provide better configuration mechanisms. 

considering this, I'd rather provide the most flexible solution (instead of the most convenient) that allows configuring any  arbitrary set of methods.

with OTEL_INSTRUMENTATION_HTTP_ADDITIONAL_KNOWN_METHODS, it won't be possible to exclude standard headers.

---

> with OTEL_INSTRUMENTATION_HTTP_ADDITIONAL_KNOWN_METHODS, it won't be possible to exclude standard methods

I think this is a good point, e.g. for users who regularly see spam across all 9 standard methods but they know their service only accepts GET and POST and want to reduce cardinality / improve bucketing of OTHER

---

I rewrote the note based on all your suggestions, thanks!